parser/macro.go

```diff
@@ -124,71 +119,69 @@ const (
-	return makeQuantifier(quantifierExists, p, loc, target, args)
+func makeExists(p *parserHelper, ctx interface{}, target *expr.Expr, args []*expr.Expr) *expr.Expr {
+	return makeQuantifier(quantifierExists, p, ctx, target, args)
 }
 
-func makeExistsOne(p *parser, loc common.Location, target ast.Expression, args []ast.Expression) ast.Expression {
-	return makeQuantifier(quantifierExistsOne, p, loc, target, args)
+func makeExistsOne(p *parserHelper, ctx interface{}, target *expr.Expr, args []*expr.Expr) *expr.Expr {
+	return makeQuantifier(quantifierExistsOne, p, ctx, target, args)
 }
 
-func makeQuantifier(kind quantifierKind, p *parser, loc common.Location, target ast.Expression, args []ast.Expression) ast.Expression {
+func makeQuantifier(kind quantifierKind, p *parserHelper, ctx interface{}, target *expr.Expr, args []*expr.Expr) *expr.Expr {
 	v, found := extractIdent(args[0])
 	if !found {
-		p.errors.argumentIsNotIdent(args[0].Location())
-		return &ast.ErrorExpression{}
+		offset := p.positions[args[0].Id]
+		location, _ := p.source.LocationFromOffset(offset)
+		return p.reportError(location, "argument must be a simple name")
 	}
-
-	accuIdent := func() ast.Expression {
-		return ast.NewIdent(p.id(), loc, accumulatorName)
+	accuIdent := func() *expr.Expr {
+		return p.newIdent(ctx, accumulatorName)
 	}
 
-	var init ast.Expression
-	var condition ast.Expression
-	var step ast.Expression
-	var result ast.Expression
+	var init *expr.Expr
+	var condition *expr.Expr
+	var step *expr.Expr
+	var result *expr.Expr
 	switch kind {
 	case quantifierAll:
-		init = ast.NewBoolConstant(p.id(), loc, true)
+		init = p.newConstBool(ctx, true)
 		condition = accuIdent()
-		step = ast.NewCallFunction(p.id(), loc, operators.LogicalAnd, accuIdent(), args[1])
+		step = p.newGlobalCall(ctx, operators.LogicalAnd, accuIdent(), args[1])
 		result = accuIdent()
 	case quantifierExists:
-		init = ast.NewBoolConstant(p.id(), loc, false)
-		condition = ast.NewCallFunction(p.id(), loc, operators.LogicalNot, accuIdent())
-		step = ast.NewCallFunction(p.id(), loc, operators.LogicalOr, accuIdent(), args[1])
+		init = p.newConstBool(ctx, false)
+		condition = p.newGlobalCall(ctx, operators.LogicalNot, accuIdent())
+		step = p.newGlobalCall(ctx, operators.LogicalOr, accuIdent(), args[1])
 		result = accuIdent()
 	case quantifierExistsOne:
-		zeroExpr := ast.NewInt64Constant(p.id(), loc, 0)
-		oneExpr := ast.NewInt64Constant(p.id(), loc, 1)
+		zeroExpr := p.newConstInt(ctx, 0)
+		oneExpr := p.newConstInt(ctx, 1)
 		init = zeroExpr
-		condition = ast.NewCallFunction(p.id(), loc, operators.LessEquals, accuIdent(), oneExpr)
-		step = ast.NewCallFunction(p.id(), loc, operators.Conditional, args[1],
-			ast.NewCallFunction(p.id(), loc, operators.Add, accuIdent(), oneExpr), accuIdent())
-		result = ast.NewCallFunction(p.id(), loc, operators.Equals, accuIdent(), oneExpr)
+		condition = p.newGlobalCall(ctx, operators.LessEquals, accuIdent(), oneExpr)
```

Comment:
It can terminate early on false, but if it encounters an error along the way it can't continue execution. It doesn't absorb errors like the exists / all.

Reply:
Right, it doesn't ignore errors, but based on our earlier discussions I got the impression that it had to do a full count of all true values, then check that the count is equal to one, and an error anywhere in the list/map caused it to error out.

If the code here is what's intended, then the semantics need to be updated.

Reply:
I've added a TODO to make this consistent with CEL semantics.